src/main/scala/org/constellation/p2p/PeerAPI.scala

```diff
@@ -143,43 +145,109 @@ class PeerAPI(override val ipManager: IPManager[IO])(
-                logger.warn(s"snapshot/info checkpointsToGet - ${checkpointsToGet}")
-
+                val checkpointsToGet = info.acceptedCBSinceSnapshotHashes.toList.diff(deSerCheckpointHashes.toList)
                 checkpointsToGet.traverse {
                   dao.checkpointService.fullData(_)
                 }.flatMap { cbs =>
                   val infoSer = info.copy(acceptedCBSinceSnapshotCache = cbs.flatten).toSnapshotInfoSer()
-                  dao.snapshotService.recentSnapshotInfo
-                    .put(ip, infoSer)
+                  val res = dao.snapshotService.recentSnapshotInfo
+                    .put(address, infoSer)
                     .flatTap(_ => IO { logger.warn(s"snapshot/info recentSnapshotInfo updating") })
                     .map(_ => Array.empty[Byte].some)
+                  res
                 }
               }
-
               APIDirective.handle {
                 dao.cluster.getNodeState
                   .map(NodeState.canActAsRedownloadSource)
                   .ifM(getInfo, IO.pure(None))
               } {
                 case None => complete(StatusCodes.NotFound)
-                case _    => complete(StatusCodes.OK, Array.empty[Byte])
+                case _    => complete(StatusCodes.OK, Array()) //need Array() here, Array.empty marshalls as List
               }
+          }
+        } ~
+          path(LongNumber) { height =>
+            entity(as[(String, Array[Array[Byte]])]) {
+              case (address, curCheckpointHashes) =>
+                val curSnap = dao.snapshotService.storedSnapshot.get.unsafeRunSync()
+                if (height == curSnap.height) {
+                  val deSerCheckpointHashes =
+                    curCheckpointHashes.flatMap(chunkDeSerialize[Seq[String]](_, "snapshot/info/curCheckpointHashes"))
+                  logger.debug(
+                    s"snapshot/info/${height} from node address: ${address} - height == curSnap.height: true  - deSerCheckpointHashes.size ${deSerCheckpointHashes.size}"
+                  )
+                  val getInfo = dao.snapshotService.getSnapshotInfo.flatMap { info =>
+                    val checkpointsToGet = info.acceptedCBSinceSnapshotHashes.toList.diff(deSerCheckpointHashes.toList)
+                    checkpointsToGet.traverse {
+                      dao.checkpointService.fullData(_)
+                    }.flatMap { cbs =>
+                      val infoSer = info.copy(acceptedCBSinceSnapshotCache = cbs.flatten).toSnapshotInfoSer()
+                      val res = dao.snapshotService.recentSnapshotInfo
+                        .put(address, infoSer)
+                        .flatTap(_ => IO { logger.warn(s"snapshot/info recentSnapshotInfo updating") })
+                        .map(_ => Array.empty[Byte].some)
+                      val testPut = dao.snapshotService.recentSnapshotInfo.lookup(address)
+                      val check = (res >> testPut).unsafeRunSync()
```

Comment:
It's a `PeerAPI`, not `PeerAPITest`. Why are you using such statements here? Additionally, `unsafeRunSync` in can lead to other bugs. We tried to get rid of all unsafe run syncs from such places because we had CPU 100% and other issues.

Reply:
That's just from testing locally